### DISCUSSION TITLE
How to speed up APISIX Development environment buildup

### DISCUSSION THREAD

**Discussion by lewiseliu** [apache/apisix#7705] [Q&A] [answered]
**Background:** 

- We are looking to contribute to APISIX or setup a development environment recently, however seems it is not easy to pass each steps on the guidance "https://github.com/apache/apisix/blob/d702d07d4bcef00eb125764e5fd1e5d101cd8d00/docs/en/latest/building-apisix.md". 
- Base Linux image: centos:centos7

**2 Issues highlighted:** 

-  - Issue 1: if without network proxy, failed installing dependency from China Mainland, when doing make deps .
-  - Issue 2: If we have network proxy, there will be some  "no file" exception, when doing apisix init. 

**Workaround Solution suggestion:**

-  - For Issue 1: seems no workaround so far except network proxy

 -  - For Issue 2:  there are 2 suggestions, 

     1) whether anybody can share the development environment docker image, so that the whole community can reuse this image and do development directly without additional buildup effort. 
     2) whether we can reuse the APISIX (docker pull apache/apisix:dev or docker pull apache/apisix:2.15.0-centos) as the development environment directly, include self-developed plugin deployment (with dockerfile RUN COPY) and testing. 

**Appendix:** 
For issue 1, the logging as below: 
```
error: RPC failed; result=7, HTTP code = 0
fatal: The remote end hung up unexpectedly
Error: Failed installing dependency: https://luarocks.org/lua-resty-healthcheck-api7-2.2.0-0.rockspec - Failed cloning git repository.
make: *** [deps] Error 1 
```

For Issue 2, we attached the logging as attachment. 
[apisix setup log--kt.log](https://github.com/apache/apisix/files/9352117/apisix.setup.log--kt.log)

**Final solution:**
1) Issue 1: use China Lua package proxy repo
```
cd apisix-${APISIX_VERSION}
make deps ENV_LUAROCKS_SERVER=https://luarocks.cn
make install

```
2) Issue 2: copy personal workdir to /usr/local/apisix
Personal workdir is /data/apisix-2.15.0, however the target environment path is expected to be /usr/local/apisix. So we need to copy all the materials to target directory: 
`# cp -R /data/apisix-2.15.0 /usr/local/apisix`

**Comment by tzssangglass** [comment, marked as answer]
> : if without network proxy, failed installing dependency from China Mainland, when doing make deps .

Mainland China users can use `luarocks.cn` as the LuaRocks server. You can use this wrapper with the Makefile to set this up:

```bash
make deps ENV_LUAROCKS_SERVER=https://luarocks.cn
```

**Reply by lewiseliu** [reply]
It works!!!  @tzssangglass we are going ahead to next step.

**Comment by tzssangglass** [comment]
> 2: If we have network proxy, there will be some "no file" exception, when doing apisix init.

This should be caused by a failure to execute make deps and not installing all the lua libs.

**Reply by lewiseliu** [reply]
After make deps ENV_LUAROCKS_SERVER=https://luarocks.cn and make install, and etcd installation.
Seems apisix init still not work well, as below: 
```
 apisix init 
/usr/local/openresty/luajit/bin/luajit /usr/local/apisix/apisix/cli/apisix.lua init
/usr/local/openresty/luajit/bin/luajit: cannot open /usr/local/apisix/apisix/cli/apisix.lua: No such file or directory

```
We will try to use Alpine as base image.

**Reply by tzssangglass** [reply]
cd apisix workdir, use `./bin/apisix init`

**Reply by lewiseliu** [reply]
Yes, it works!!!  @tzssangglass  you are Hero!!!
Per the guidance, the final workdir should be /..../apisix-2.15.0/. We are confused why the binary file was not installed on envirnoment path by default. **If so, we suggest to update the guidenace to highlight this workdir point.** 

Below is our operation log:
```
# pwd
/data/apisix-2.15.0
# bin/apisix init
/usr/local/openresty/luajit/bin/luajit ./apisix/cli/apisix.lua init
# ps -ef|grep apisix
root       13203   13161  0 03:54 pts/1    00:00:00 grep --color=auto apisix
# mkdir logs
# bin/apisix test
/usr/local/openresty/luajit/bin/luajit ./apisix/cli/apisix.lua test
configuration test is successful
# ./bin/apisix start
/usr/local/openresty/luajit/bin/luajit ./apisix/cli/apisix.lua start
# ps -ef|grep apisix
root       13275       1  0 03:56 ?        00:00:00 nginx: master process /usr/bin/openresty -p /data/apisix-2.15.0 -c /data/apisix-2.15.0/conf/nginx.conf
root       13285   13161  0 03:56 pts/1    00:00:00 grep --color=auto apisix
# ./bin/apisix stop
/usr/local/openresty/luajit/bin/luajit ./apisix/cli/apisix.lua stop
```

**Comment by tzssangglass** [comment]
> whether anybody can share the development environment docker image, so that the whole community can reuse this image and do development directly without additional buildup effort.

I guess mirroring is not enough to support a development environment? Anyway, hear what others think.

**Reply by lewiseliu** [reply]
With assistance from @tzssangglass, APISIX DEVELOP IMAGE was built-up and released on docker hub, sharing for community, as: 
lewiseliu/apisix-dev-env-sshd:2.15.0 (come with sshd, hub.docker.com)`
Readme for this image:
```
docker pull lewiseliu/apisix-dev-env-sshd:2.15.0
docker run -itd -p 2222:22 --name apisix-dev-ssh  {$image_id}  /usr/sbin/sshd -D
access check from VS Code PC, e.g. ssh root@192.168.226.129(docker host machine) -p 2222
```

**Reply by tokers** [reply]
@lewiseliu @tzssangglass I think we may add the `luarocks.cn` to the docs?

**Reply by tzssangglass** [reply]
FAQ.md has this

**Comment by Chever-John** [comment]
So, why don't we state it in the documentation?

`make deps ENV_LUAROCKS_SERVER=https://luarocks.cn`

in this [doc link](https://apisix.apache.org/docs/apisix/next/building-apisix/)

**Comment by lewiseliu** [comment]
@tzssangglass We are simulating the Unit test scenario with one existing test code (e.g. t/plugin/jwt-auth.t), and we attempted to practice below steps from 3rd party website:
PATH=/usr/local/openresty/nginx/sbin:/usr/bin PERL5LIB=.:$PERL5LIB FLUSH_ETCD=1 prove -Itest-nginx/lib -r t/admin
However, seems it can not running well, any suggestion?

**Reply by lewiseliu** [reply]
As LTS version, Seems official documentation for V2.13 is more comprehensive. We found some test operation on https://apisix.apache.org/zh/docs/apisix/2.13/how-to-build/ , then we will follow up this guide in advance.

**Reply by tzssangglass** [reply]
Our documentation team is refactoring the documentation. I found that we are currently missing detailed step-by-step documentation for building the development environment from the initialization environment in the source code. I was thinking that maybe I could provide a video tutorial to show how I started building a development environment from a clean CentOS 7 environment. (But eventually, we'll need a document)

**Reply by lewiseliu** [reply]
I am trying to record every detail steps as successful reference, and I am glad to share even though I am still struggle on the progress.
